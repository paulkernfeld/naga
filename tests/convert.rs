fn load_wgsl(name: &str) -> naga::Module {
    let path = format!("{}/test-data/{}.wgsl", env!("CARGO_MANIFEST_DIR"), name);
    let input = std::fs::read_to_string(path).unwrap();
    naga::front::wgsl::parse_str(&input).unwrap()
}

#[test]
fn convert_quad() {
    let module = load_wgsl("quad");
    {
        use naga::back::msl;
        let mut binding_map = msl::BindingMap::default();
        binding_map.insert(
            msl::BindSource { set: 0, binding: 0 },
            msl::BindTarget { buffer: None, texture: Some(1), sampler: None, mutable: false },
        );
        binding_map.insert(
            msl::BindSource { set: 0, binding: 1 },
            msl::BindTarget { buffer: None, texture: None, sampler: Some(1), mutable: false },
        );
        let options = msl::Options {
            binding_map: &binding_map,
        };
        msl::write_string(&module, options).unwrap();
    }
}

#[test]
fn convert_boids() {
    let module = load_wgsl("boids");
    {
        use naga::back::msl;
        let mut binding_map = msl::BindingMap::default();
        binding_map.insert(
            msl::BindSource { set: 0, binding: 0 },
            msl::BindTarget { buffer: Some(0), texture: None, sampler: None, mutable: false },
        );
        binding_map.insert(
            msl::BindSource { set: 0, binding: 1 },
            msl::BindTarget { buffer: Some(1), texture: None, sampler: Some(1), mutable: false },
        );
        binding_map.insert(
            msl::BindSource { set: 0, binding: 2 },
            msl::BindTarget { buffer: Some(2), texture: None, sampler: Some(1), mutable: false },
        );
        let options = msl::Options {
            binding_map: &binding_map,
        };
        msl::write_string(&module, options).unwrap();
    }
}

#[test]
#[cfg(all(feature = "serialize", feature = "deserialize"))]
fn boids_serde_round_trip() {
    let module = load_wgsl("boids");
    let module_ser = ron::ser::to_string(&module).unwrap();
    let module_de = ron::de::from_str::<naga::Module>(&module_ser).unwrap();
    let module_ser_ser = ron::ser::to_string(&module_de).unwrap();
    assert_eq!(module_ser, module_ser_ser);
}
