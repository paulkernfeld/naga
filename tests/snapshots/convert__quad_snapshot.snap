---
source: tests/convert.rs
expression: "load_wgsl(\"quad\")"
---
Module(
  header: Header(
    version: (1, 0, 0),
    generator: 0,
  ),
  types: Arena(
    data: [
      Type(
        name: None,
        inner: Scalar(
          kind: Float,
          width: 32,
        ),
      ),
      Type(
        name: None,
        inner: Vector(
          size: Bi,
          kind: Float,
          width: 32,
        ),
      ),
      Type(
        name: None,
        inner: Vector(
          size: Quad,
          kind: Float,
          width: 32,
        ),
      ),
      Type(
        name: None,
        inner: Scalar(
          kind: Sint,
          width: 32,
        ),
      ),
    ],
  ),
  constants: Arena(
    data: [
      Constant(
        name: Some("c_scale"),
        specialization: None,
        inner: Float(1.2),
        ty: Handle(1),
      ),
      Constant(
        name: None,
        specialization: None,
        inner: Float(0),
        ty: Handle(1),
      ),
      Constant(
        name: None,
        specialization: None,
        inner: Float(1),
        ty: Handle(1),
      ),
      Constant(
        name: None,
        specialization: None,
        inner: Sint(1),
        ty: Handle(4),
      ),
      Constant(
        name: None,
        specialization: None,
        inner: Sint(0),
        ty: Handle(4),
      ),
    ],
  ),
  global_variables: Arena(
    data: [
      GlobalVariable(
        name: Some("a_pos"),
        class: Input,
        binding: Some(Location(0)),
        ty: Handle(2),
      ),
      GlobalVariable(
        name: Some("a_uv"),
        class: Input,
        binding: Some(Location(1)),
        ty: Handle(2),
      ),
      GlobalVariable(
        name: Some("v_uv"),
        class: Output,
        binding: Some(Location(0)),
        ty: Handle(2),
      ),
      GlobalVariable(
        name: Some("o_position"),
        class: Output,
        binding: Some(BuiltIn(Position)),
        ty: Handle(3),
      ),
      GlobalVariable(
        name: Some("a_uv"),
        class: Input,
        binding: Some(Location(0)),
        ty: Handle(2),
      ),
      GlobalVariable(
        name: Some("o_color"),
        class: Output,
        binding: Some(Location(0)),
        ty: Handle(3),
      ),
    ],
  ),
  functions: Arena(
    data: [
      Function(
        name: Some("main_vert"),
        control: FunctionControl(
          bits: 0,
        ),
        parameter_types: [],
        return_type: None,
        global_usage: [
          GlobalUse(
            bits: 1,
          ),
          GlobalUse(
            bits: 0,
          ),
          GlobalUse(
            bits: 0,
          ),
          GlobalUse(
            bits: 2,
          ),
        ],
        local_variables: Arena(
          data: [],
        ),
        expressions: Arena(
          data: [
            Constant(Handle(1)),
            GlobalVariable(Handle(1)),
            GlobalVariable(Handle(3)),
            GlobalVariable(Handle(4)),
            GlobalVariable(Handle(2)),
            Binary(
              op: Multiply,
              left: Handle(1),
              right: Handle(2),
            ),
            Constant(Handle(2)),
            Constant(Handle(3)),
            Compose(
              ty: Handle(3),
              components: [
                Handle(6),
                Handle(7),
                Handle(8),
              ],
            ),
          ],
        ),
        body: [
          Store(
            pointer: Handle(4),
            value: Handle(9),
          ),
          Return(
            value: None,
          ),
        ],
      ),
      Function(
        name: Some("main_frag"),
        control: FunctionControl(
          bits: 0,
        ),
        parameter_types: [],
        return_type: None,
        global_usage: [
          GlobalUse(
            bits: 0,
          ),
          GlobalUse(
            bits: 0,
          ),
          GlobalUse(
            bits: 0,
          ),
          GlobalUse(
            bits: 0,
          ),
          GlobalUse(
            bits: 0,
          ),
          GlobalUse(
            bits: 2,
          ),
        ],
        local_variables: Arena(
          data: [],
        ),
        expressions: Arena(
          data: [
            Constant(Handle(1)),
            GlobalVariable(Handle(1)),
            GlobalVariable(Handle(3)),
            GlobalVariable(Handle(4)),
            GlobalVariable(Handle(6)),
            GlobalVariable(Handle(5)),
            Constant(Handle(4)),
            Constant(Handle(5)),
            Constant(Handle(5)),
            Constant(Handle(4)),
            Compose(
              ty: Handle(3),
              components: [
                Handle(7),
                Handle(8),
                Handle(9),
                Handle(10),
              ],
            ),
          ],
        ),
        body: [
          Store(
            pointer: Handle(5),
            value: Handle(11),
          ),
          Return(
            value: None,
          ),
        ],
      ),
    ],
  ),
  entry_points: [
    EntryPoint(
      exec_model: Vertex,
      name: "main",
      function: Handle(1),
    ),
    EntryPoint(
      exec_model: Fragment,
      name: "main",
      function: Handle(2),
    ),
  ],
)
